### PR TITLE
Use a single cib__operation_t table for pacemaker-based and libcib

### PR DESCRIPTION
**Pull request by nrwahl2**
This allows us to deduplicate the shared data (op name and op attributes). We introduce a new `enum cib__op_type` to use as a key for accessing `cib_op_fn_t` functions. This is a step toward a cleaner CIB transaction implementation, which is coming soon.

We drop `cib_operation_t:prepare` and `cib_operation_t:cleanup` as well.

I was going to go farther with the XML stuff but dropped that for now. I included in this PR what I had already done. Getting rid of `getDocPtr()` doesn't actually allow all that many switches to const (I only checked `xml.c` so I'm sure I missed a bunch that we can update later). Most of the barriers to const come from libxml2 itself.

### CODE REVIEW

**Review by kgaillot — COMMENTED** [review]
looking good

File: lib/common/output_html.c [hunk 1]
```diff
@@ -152,7 +152,7 @@ html_finish(pcmk__output_t *out, crm_exit_t exit_status, bool print, void **copy
      * anything else that the user could add, and we want it done last to pick up
      * any options that may have been given.
      */
-    head_node = xmlNewNode(NULL, (pcmkXmlStr) "head");
+    head_node = xmlNewDocRawNode(NULL, NULL, (pcmkXmlStr) "head", NULL);
```
[the remark this answers is not in the document]
Comment (nrwahl2) [reply lib/common/output_html.c:155]: Wouldn't hurt if we wanted to use the non-raw version here. Currently we use the raw version in `create_xml_node()` and `pcmk__expand_tags_in_sets()`, and we don't use the non-raw version anywhere.

---

More detail: Right, but we don't need entities support for this particular call. Entities go into the `content` argument (the fourth argument of `xmlNewDocNode()` or `xmlNewDocRawNode()`). Here, `content` is `NULL`.

https://gnome.pages.gitlab.gnome.org/libxml2/devhelp/libxml2-tree.html#xmlNewDocNode

You can also see the implementations: `xmlNewDocRawNode()` calls `xmlNewDocNode()` with `content == NULL` and then sets children to `content` if not `NULL`. `content` is already `NULL`, so it should be all the same.
https://gitlab.gnome.org/GNOME/libxml2/-/blob/master/tree.c#L2408-2434
https://gitlab.gnome.org/GNOME/libxml2/-/blob/master/tree.c#L2332-2367

File: include/crm/common/xml.h [hunk 2]
```diff
@@ -52,8 +52,6 @@ typedef const xmlChar *pcmkXmlStr;
 gboolean add_message_xml(xmlNode * msg, const char *field, xmlNode * xml);
 xmlNode *get_message_xml(const xmlNode *msg, const char *field);
 
-xmlDoc *getDocPtr(xmlNode * node);
```
[the remark this answers is not in the document]
Comment (nrwahl2) [reply include/crm/common/xml.h:55]: ah crap, good catch. Most or all of the public XML library code ought to be deprecated soon.

File: daemons/based/based_callbacks.c [hunk 3]
```diff
@@ -1458,11 +1507,7 @@ cib_process_command(xmlNode *request, const cib_operation_t *operation,
         goto done;
     }
 
-    rc = operation->prepare(request, &input, &section);
-    if (rc != pcmk_ok) {
-        crm_trace("Failed to prepare operation: %s", pcmk_strerror(rc));
-        goto done;
-    }
+    input = prepare_input(request, operation->type);
```
[the remark this answers is not in the document]
Comment (nrwahl2) [reply daemons/based/based_callbacks.c:1519]: Ouch, that's quite an oversight, surprised it didn't break any tests. I'll pass it out via a pointer argument.

File: lib/common/output_html.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/common/output_html.c:155]: great

**Comment by nrwahl2** [comment]
Addressed review: updated commits, and added a new one on top to check `xml->doc != NULL`. I limited it to `schemas.c` -- if needed I can do it everywhere that we check `xml != NULL` even if we don't directly use `xml->doc`.

File: daemons/based/based_callbacks.c [hunk 4]
```diff
@@ -1278,6 +1278,57 @@ cib_process_request(xmlNode *request, gboolean privileged,
...
+        return NULL;
+    }
+
+    if (pcmk__str_any_of(crm_element_name(root), F_CRM_DATA, F_CIB_CALLDATA,
+                         NULL)) {
+        root = first_named_child(root, XML_TAG_CIB);
+    }
+
+    // Grab the specified section
+    if ((root != NULL) && (section != NULL)
```
Comment (kgaillot) [inline daemons/based/based_callbacks.c:1331]: *section

Comment (nrwahl2) [reply daemons/based/based_callbacks.c]: d'oh

**Review by kgaillot — APPROVED** [review]

**Pull request merged**